File: LeetCode/0918. Maximum Sum Circular Subarray/solution.py
"""
111 / 111 test cases passed.
Runtime: 188 ms
Memory Usage: 18.4 MB
"""
class Solution:
    def maxSubarraySumCircular(self, nums: List[int]) -> int:
        max_sum = min_sum = nums[0]
        total = maxn = minn = 0
        for num in nums:
            maxn = max(maxn + num, num)
            max_sum = max(max_sum, maxn)
            minn = min(minn + num, num)
            min_sum = min(min_sum, minn)
            total += num
        return max(max_sum, total - min_sum) if max_sum > 0 else max_sum

File: LeetCode/0918. Maximum Sum Circular Subarray/solution.cpp
/**
 * 111 / 111 test cases passed.
 * Runtime: 72 ms
 * Memory Usage: 38.9 MB 
 */
class Solution {
public:
    int maxSubarraySumCircular(vector<int>& nums) {
        int total = 0, maxSum = nums[0], maxn = 0, minSum = nums[0], minn = 0;
        for (auto &num : nums) {
            maxn = max(maxn + num, num);
            maxSum = max(maxSum, maxn);
            minn = min(minn + num, num);
            minSum = min(minSum, minn);
            total += num;
        }
        return maxSum > 0 ? max(maxSum, total - minSum) : maxSum;
    }
};
